### PR TITLE
alert manager compatibility 

### PR DESCRIPTION
**Pull request by ydye**
Old
```
prometheus:
  alerting:
      alert_manager_port: 9093
      alert_receiver: alert@example.com
      smtp_url: smtp.gmail.com:587
      smtp_from: foo_bar@gmail.com
      smtp_auth_username: user@gmail.com
      smtp_auth_password: gmail_password
```

New
```
alert-manager:
   receiver: your_addr@example.com
   smtp_url: smtp.office365.com:587
   smtp_from: alert_sender@example.com
   smtp_auth_username: alert_sender@example.com
   smtp_auth_password: password_for_alert_sender
   port: 9093 
```

### CODE REVIEW

File: deployment/clusterObjectModel/forward_compatibility.py [hunk 1]
```diff
@@ -65,11 +65,19 @@ def service_configuration_convert(service_configuration):
             service_configuration["hadoop-resource-manager"]["yarn_exporter_port"] = service_configuration["prometheus"]["yarn_exporter_port"]
         if "prometheus-port" in service_configuration["prometheus"] and "port" not in service_configuration["prometheus"]:
             service_configuration["prometheus"]["port"] = service_configuration["prometheus"]["prometheus-port"]
+        if "alerting" in service_configuration["prometheus"]:
+            service_configuration["alert-manager"] = service_configuration["prometheus"]["alerting"]
```
Comment (xudifsd) [inline deployment/clusterObjectModel/forward_compatibility.py]: does alert-manager always exist?

Comment (ydye) [reply deployment/clusterObjectModel/forward_compatibility.py]: Depend on whether ```alerting``` exits or not.

Comment (ydye) [reply deployment/clusterObjectModel/forward_compatibility.py]: Changed

Comment (xudifsd) [reply deployment/clusterObjectModel/forward_compatibility.py]: With so many ifs, I think maybe you should implement a function that can do this.

Maybe `transform(old_model, "prometheus.alert.alert_manager_port", new_model, "alert-manager.receiver")` which means copy one field from one object to another field of another object, and handle this no-exist field problem in the transform function. This way it's more easy to debug and find bugs.

File: deployment/clusterObjectModel/forward_compatibility.py [hunk 2]
```diff
@@ -17,59 +17,76 @@
...
+def transform(old_model, old_key, new_model, new_key):
+    old_key_list = old_key.split('.')
+    new_key_list = new_key.split('.')
+    old_dict = old_model
+    new_dict = new_model
+    for key in old_key_list:
+        if key not in old_dict:
+            return
+        old_dict = old_dict[key]
+    for key in new_key_list:
```
Comment (xudifsd) [inline deployment/clusterObjectModel/forward_compatibility.py]: looks like you only need to iterate over `new_key_list[:-1]` and remove the check of `!= new_key_list[-1]`

File: deployment/clusterObjectModel/forward_compatibility.py [hunk 3]
```diff
@@ -17,59 +17,76 @@
...
+        if key not in new_model and key != new_key_list[-1]:
+            new_dict[key] = dict()
+        if key != new_key_list[-1]
+            new_dict = new_dict[key]
+    new_dict[new_key_list[-1]] = old_dict
```
Comment (xudifsd) [inline deployment/clusterObjectModel/forward_compatibility.py]: Maybe write some unit test to test edge cases.

File: deployment/clusterObjectModel/forward_compatibility.py (hunk 2 above)
Comment (ydye) [reply deployment/clusterObjectModel/forward_compatibility.py]: Done

File: deployment/clusterObjectModel/forward_compatibility.py [hunk 4]
```diff
@@ -17,59 +17,76 @@
...
-        if "prometheus-port" in service_configuration["prometheus"] and "port" not in service_configuration["prometheus"]:
-            service_configuration["prometheus"]["port"] = service_configuration["prometheus"]["prometheus-port"]
-
-    if "grafana" in service_configuration:
-        if "grafana-port" in service_configuration["grafana"] and "port" not in service_configuration["grafana"]:
-            service_configuration["grafana"]["port"] = service_configuration["grafana"]["grafana-port"]
-
-    return service_configuration
+
+    new_configuration = dict()
```
[the remark this answers is not in the document]
Comment (ydye) [reply deployment/clusterObjectModel/forward_compatibility.py]: Done

File: deployment/clusterObjectModel/forward_compatibility.py [hunk 5]
```diff
@@ -17,59 +17,75 @@
...
+    new_key_list = new_key.split('.')
+    old_dict = old_model
+    new_dict = new_model
+    for key in old_key_list:
+        if key not in old_dict:
+            return
+        old_dict = old_dict[key]
+    for key in new_key_list[:-1]:
+        if key not in new_model and key != new_key_list[-1]:
+            new_dict[key] = dict()
```
[the remark this answers is not in the document]
Comment (ydye) [reply deployment/clusterObjectModel/forward_compatibility.py]: Done

File: deployment/clusterObjectModel/forward_compatibility.py [hunk 6]
```diff
@@ -17,59 +17,75 @@
...
+    old_key_list = old_key.split('.')
+    new_key_list = new_key.split('.')
+    old_dict = old_model
+    new_dict = new_model
+    for key in old_key_list:
+        if key not in old_dict:
+            return
+        old_dict = old_dict[key]
+    for key in new_key_list[:-1]:
+        if key not in new_model and key != new_key_list[-1]:
```
[the remark this answers is not in the document]
Comment (ydye) [reply deployment/clusterObjectModel/forward_compatibility.py]: Done
